### PR TITLE
feat(WASM): implement the PreviewKeyDown/Up event and handler in WASM

### PR DESCRIPTION
**Pull request by TopProgrammer77**
GitHub Issue (If applicable): closes #

## PR Type

What kind of change does this PR introduce?

- Feature
## What is the current behavior?

 KeyDown & KeyUp events are supported on WASM
PreviewKeyDown/Up events aren't supported.

## What is the new behavior?

PreviewKeyDown/Up events are supported on WASM.

### <samp>🤖 Generated by Copilot at e1aaf31</samp>

This pull request adds support for the capture phase of event propagation and the preview key down and key up events for the WASM platform. It modifies the `WindowManager` and `UIElement` classes to handle these features, and adds a new test class and a sample page to demonstrate them.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

https://github.com/unoplatform/uno/issues/1595
Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs [hunk 1]
```diff
@@ -434,6 +438,22 @@ private static object CoerceRoutedEventFlag(DependencyObject dependencyObject, o
...
+		}
+#endif
+
+#if __WASM__
+		public event KeyEventHandler PreviewKeyUp
+		{
+			add => AddHandler(PreviewKeyUpEvent, value, false);
+			remove => RemoveHandler(PreviewKeyUpEvent, value);
+		}
+#endif
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs:455]: This could be a single `#if __WASM__` rathern than two

Comment (TopProgrammer77) [reply src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs]: Thank you, @Youssef1313

**Comment by TopProgrammer77** [comment]
Hi, @Youssef1313 , @jeromelaban , the test(When_Keyboard_Up_And_Down) is passing on my side, so I am not sure how to fix this. Could you give me some advise?

**Review by Youssef1313 — COMMENTED** [review]
https://github.com/unoplatform/uno/blob/master/src/SourceGenerators/Uno.UI.SourceGenerators/ImplementedRoutedEvents/ImplementedRoutedEventsGenerator.cs

@MartinZikmund Would we need to update the generator here?

**Comment by TopProgrammer77** [comment]
Hi, @dr1rrb. Could you review my code and give some advice?

**Comment by DierkDroth** [comment]
@dr1rrb @Youssef1313 is there any chance to get this PR reviewed and merged? I would be happy to test a new build then...

**Review by dr1rrb — COMMENTED** [review]
You will also have to move/implementd the overridable methods `Control.OnPreviewKey<Down|Up>` from https://github.com/unoplatform/uno/blob/master/src/Uno.UI/Generated/3.0.0.0/Windows.UI.Xaml.Controls/Control.cs#L314. You will then have to add support of those methods here https://github.com/unoplatform/uno/blob/master/src/SourceGenerators/Uno.UI.SourceGenerators/ImplementedRoutedEvents/ImplementedRoutedEventsGenerator.cs#L48

**Comment by dr1rrb** [comment]
Also, I realized that as this event is a tunneling event, we need to make sure that the event will never bubble in managed code. To do this I think that adding a simple [`IsTunnelingEvent` here](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/RoutedEvent.cs#L50), set it to for the `PreviewKey<Down|Up>Event` and then update the [condition here](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs#L702) to validate that flag like `if (routedEvent.IsTunnelingEvent || ctx.Mode.HasFlag(BubblingMode.IgnoreParents) || ctx.Root == this)` would do be enough.

**Review by dr1rrb — APPROVED** [review]

**Comment by TopProgrammer77** [comment]
@dr1rrb , thanks for your help.

**Pull request merged**